kong/mlcache.lua

```diff
@@ -0,0 +1,466 @@
+        if type(opts) ~= "table" then
+            return error("opts must be a table")
+        end
+
+        if opts.ttl then
+            if type(opts.ttl) ~= "number" then
+                return error("opts.ttl must be a number")
+            end
+
+            if opts.ttl < 0 then
+                return error("opts.ttl must be >= 0")
+            end
+        end
+
+        if opts.neg_ttl then
+            if type(opts.neg_ttl) ~= "number" then
+                return error("opts.neg_ttl must be a number")
+            end
+
+            if opts.neg_ttl < 0 then
+                return error("opts.neg_ttl must be >= 0")
+            end
+        end
+
+        ttl     = opts.ttl
+        neg_ttl = opts.neg_ttl
+    end
+
+    if not ttl then
+        ttl = self.ttl
+    end
+
+    if not neg_ttl then
+        neg_ttl = self.neg_ttl
+    end
+
+    -- worker LRU cache retrieval
+
+    local data = self.lru:get(key)
+    if data == CACHE_MISS_SENTINEL_LRU then
+        return nil
+    end
+
+    if data ~= nil then
+        return data
+    end
+
+    -- not in worker's LRU cache, need shm lookup
+
+    local err
+    data, err = shmlru_get(self, key)
+    if err then
+        return nil, err
+    end
+
+    if data == CACHE_MISS_SENTINEL_LRU then
+        return nil
+    end
+
+    if data ~= nil then
+        return data
+    end
+
+    -- not in shm either
+    -- single worker must execute the callback
+
+    local lock, err = resty_lock:new(self.shm)
+    if not lock then
+        return nil, "could not create lock: " .. err
+    end
+
+    local elapsed, err = lock:lock(LOCK_KEY_PREFIX .. key)
+    if not elapsed then
+        return nil, "could not aquire callback lock: " .. err
+    end
+
+    -- check for another worker's success at running the callback
+
+    data, err = shmlru_get(self, key)
+    if err then
+        return unlock_and_ret(lock, nil, err)
+    end
+
+    if data then
+        if data == CACHE_MISS_SENTINEL_LRU then
+            return unlock_and_ret(lock, nil)
+        end
+
+        return unlock_and_ret(lock, data)
+    end
+
+    -- still not in shm, we are responsible for running the callback
+
+    local ok, err = pcall(cb, ...)
+    if not ok then
+        return unlock_and_ret(lock, nil, "callback threw an error: " .. err)
+    end
+
+    local value, err = shmlru_set(self, key, err, ttl, neg_ttl)
+    if err then
+        return unlock_and_ret(lock, nil, err)
+    end
+
+    return unlock_and_ret(lock, value)
+end
+
+
+function _M:probe(key)
+    if type(key) ~= "string" then
+        return error("key must be a string")
+    end
+
+    local v, err = self.dict:get(key)
+    if err then
+        return nil, "could not read from lua_shared_dict: " .. err
+    end
+
+    if v ~= nil then
+        local str_serialized, value_type, at, ttl = unmarshallers.shm_value(v)
+
+        local remaining_ttl = ttl - (now() - at)
+
+        -- value_type of 0 is a nil entry
+        if value_type == 0 then
```

Comment:
i wonder if we should add an unmarshaller for nil types, so we avoid a branch here? such a dispatch could simply just return `nil` itself, so we would return the return signature of `remaining_ttl, nil, nil`. my guess is we'll have quite a lot of negative caches, so this branch becomes fairly unbiased (though this is a fairly subjective assumption).